ast/file_info.go

```diff
@@ -144,101 +143,206 @@ func (f *FileInfo) AddComment(comment, attributedTo Token) Comment {
+	// in f.comments? Uh oh... that shouldn't be possible.
+	return TokenError, Comment{}
 }
 
 func (f *FileInfo) isDummyFile() bool {
-	return f.lines == nil
+	return f == nil || f.lines == nil
+}
+
+// Sequence represents a navigable sequence of elements.
+type Sequence[T any] interface {
+	// First returns the first element in the sequence. The bool return
+	// is false if this sequence contains no elements. For example, an
+	// empty file has no items or tokens.
+	First() (T, bool)
+	// Next returns the next element in the sequence that comes after
+	// the given element. The bool returns is false if there is no next
+	// item (i.e. the given element is the last one). It also returns
+	// false if the given element is invalid.
+	Next(T) (T, bool)
+	// Last returns the last element in the sequence. The bool return
+	// is false if this sequence contains no elements. For example, an
+	// empty file has no items or tokens.
+	Last() (T, bool)
+	// Previous returns the previous element in the sequence that comes
+	// before the given element. The bool returns is false if there is no
+	// previous item (i.e. the given element is the first one). It also
+	// returns false if the given element is invalid.
+	Previous(T) (T, bool)
+}
+
+func (f *FileInfo) Items() Sequence[Item] {
+	return items{fileInfo: f}
+}
+
+func (f *FileInfo) Tokens() Sequence[Token] {
+	return tokens{fileInfo: f}
 }
 
-// FirstToken returns the first token in the file. It returns
-// TokenError if the file contains no tokens (e.g. it is empty
-// or contains only whitespace and comments).
-func (f *FileInfo) FirstToken() Token {
-	return f.tokenForward(0)
+type items struct {
+	fileInfo *FileInfo
 }
 
-// NextToken returns the next token in the file that comes after
-// t. It returns TokenError if there is no next token (i.e. the
-// given t is the last token). It also returns TokenError if
-// the given t is invalid.
-func (f *FileInfo) NextToken(t Token) Token {
-	if t < 0 || int(t) >= len(f.tokens) {
-		return TokenError
+func (i items) First() (Item, bool) {
+	if len(i.fileInfo.items) == 0 {
+		return 0, false
 	}
-	return f.tokenForward(t + 1)
+	return 0, true
 }
 
-func (f *FileInfo) tokenForward(t Token) Token {
-	end := Token(len(f.tokens))
-	for t < end {
-		// Comments are not tokens in the exported API. But they
-		// do mingle with them inside of `f.tokens`, as a matter
-		// of convenience since they share the same needs for
-		// location book-keeping.
-		if !f.isComment(t) {
-			return t
-		}
-		t++
+func (i items) Next(item Item) (Item, bool) {
+	if item < 0 || int(item) >= len(i.fileInfo.items) {
+		return 0, false
+	}
+	return i.fileInfo.itemForward(item+1, true)
+}
+
+func (i items) Last() (Item, bool) {
+	if len(i.fileInfo.items) == 0 {
+		return 0, false
+	}
+	return Item(len(i.fileInfo.items) - 1), true
+}
+
+func (i items) Previous(item Item) (Item, bool) {
+	if item < 0 || int(item) >= len(i.fileInfo.items) {
```

Comment:
```suggestion
	if item <= 0 || int(item) >= len(i.fileInfo.items) {
```

Reply:
Done. I also updated the check in the `Next` implementations similarly (to early out if the item `== len(i.fileInfo.items)-1`.